controlplane/src/core/workers/DeleteUserQueue.ts

```diff
@@ -0,0 +1,141 @@
+import { UserRepository } from '../repositories/UserRepository.js';
+import Keycloak from '../services/Keycloak.js';
+import { PlatformWebhookService } from '../webhooks/PlatformWebhookService.js';
+
+const QueueName = 'user.delete';
+const WorkerName = 'DeleteUserWorker';
+
+export interface DeleteUserInput {
+  userId: string;
+}
+
+export class DeleteUserQueue {
+  private readonly queue: Queue<DeleteUserInput>;
+  private readonly logger: pino.Logger;
+
+  constructor(log: pino.Logger, conn: ConnectionOptions) {
+    this.logger = log.child({ queue: QueueName });
+    this.queue = new Queue<DeleteUserInput>(QueueName, {
+      connection: conn,
+      defaultJobOptions: {
+        removeOnComplete: true,
+        removeOnFail: true,
+        attempts: 3,
+        backoff: {
+          type: 'exponential',
+          delay: 10_000,
+        },
+      },
+    });
+
+    this.queue.on('error', (err) => {
+      this.logger.error(err, 'Queue error');
+    });
+  }
+
+  public addJob(job: DeleteUserInput, opts?: Omit<JobsOptions, 'jobId'>) {
+    return this.queue.add(job.userId, job, {
+      ...opts,
+      jobId: job.userId,
+    });
+  }
+
+  public removeJob(job: DeleteUserInput) {
+    return this.queue.remove(job.userId);
+  }
+
+  public getJob(job: DeleteUserInput) {
+    return this.queue.getJob(job.userId);
+  }
+}
+
+class DeleteUserWorker {
+  constructor(
+    private input: {
+      redisConnection: ConnectionOptions;
+      db: PostgresJsDatabase<typeof schema>;
+      logger: pino.Logger;
+      keycloakClient: Keycloak;
+      keycloakRealm: string;
+      blobStorage: BlobStorage;
+      platformWebhooks: PlatformWebhookService;
+    },
+  ) {
+    this.input.logger = input.logger.child({ worker: WorkerName });
+  }
+
+  public async handler(job: Job<DeleteUserInput>) {
+    try {
+      const userRepo = new UserRepository(this.input.logger, this.input.db);
+      const orgRepo = new OrganizationRepository(this.input.logger, this.input.db);
+
+      await this.input.keycloakClient.authenticateClient();
+
+      const user = await userRepo.byId(job.data.userId);
+      if (!user || !user.id || !user.email) {
+        throw new Error('User not found');
+      }
+
+      const { isSafe, soloOrganizations, unsafeOrganizations } = await orgRepo.canUserBeDeleted(job.data.userId);
+
+      if (!isSafe) {
+        this.input.logger.error(
+          {
+            userId: job.data.userId,
+            soloOrganizations,
+            unsafeOrganizations,
+          },
+          'Unsafe to delete user',
+        );
+        return;
+      }
+
+      await userRepo.deleteUser(
+        {
+          id: job.data.userId,
+          keycloakClient: this.input.keycloakClient,
+          keycloakRealm: this.input.keycloakRealm,
+        },
+        this.input.blobStorage,
+      );
+
+      this.input.platformWebhooks.send(PlatformEventName.USER_DELETE_SUCCESS, {
```

Comment:
This is dangerous. This will send the mail without waiting for delivery correct? In case of an error, the job is not marked as failed and in case of a shutdown, the job might not be done in time.

Reply:
Tha api is designed to not be blocking. It should be respected here as well, We log errors for these within the webhook service for which we can trigger manualy.